File: src/gmp_test/src/simulation/nDoF_gmp_test.m
%% Simulates a GMP

clc;
close all;
clear;

%% =============  includes...  =============
addpath('../../../matlab/lib/gmp_lib/');
import_gmp_lib();

addpath('../../../matlab/lib/io_lib/');
import_io_lib();


%% =============  Load params  =============
train_method = 'LS';
N_kernels = 25;
kernels_std_scaling = 1.5;

train_filename = 'data/pos_data.bin';

scale_type = 'rot_wb'; % {"prop", "rot_min", "rot_wb"}
wb_normal = [0; 0; 1]; % work-bench normal

spat_s = [2; 3; 1.1]; % spatial scale
temp_s = 1.2; % temporal scale

read_gmp_from_file = false;
write_gmp_to_file = false;

gmp_filename = 'data/gmp_pos.bin';
          

%% =============  Load train data  =============
fid = FileIO(train_filename, FileIO.in );
Timed = fid.read('Timed');
Pd_data = fid.read('Pd_data');
dPd_data = fid.read('dPd_data');
ddPd_data = fid.read('ddPd_data');
fid.close();

Ts = Timed(2) - Timed(1);


%% =============  Create/Train GMP  =============

n_dof = size(Pd_data, 1);

if (read_gmp_from_file)
    
    gmp = GMP_nDoF();
    gmp_.read(gmp, gmp_filename, '');

else
    
    %% initialize and train GMP
    gmp = GMP_nDoF(n_dof, N_kernels, kernels_std_scaling);
    tic
    offline_train_mse = gmp.train(train_method, Timed/Timed(end), Pd_data);
    offline_train_mse
    toc

    % traj_sc = TrajScale_Prop(n_dof);
    % traj_sc = TrajScale_Rot_min();
    traj_sc = TrajScale_Rot_wb();
    traj_sc.setWorkBenchNormal([0; 0; 1]);

    gmp.setScaleMethod(traj_sc);

end

% set scaling type
if ( strcmpi(scale_type, 'prop') ), traj_sc = TrajScale_Prop(n_dof);
elseif ( strcmpi(scale_type, 'rot_min') ), traj_sc = TrajScale_Rot_min();
elseif ( strcmpi(scale_type, 'rot_wb') )
    traj_sc = TrajScale_Rot_wb();
    traj_sc.setWorkBenchNormal(wb_normal);
else, error(['Unsupported scale type ''' scale_type '''...\n']);
end

gmp.setScaleMethod(traj_sc);
  

if (write_gmp_to_file), gmp_.write(gmp, gmp_filename, ''); end

%% DMP simulation
disp('GMP simulation...');
tic

P0d = Pd_data(:,1);
Pgd = Pd_data(:,end);
P0 = P0d;
Pg = spat_s.*(Pgd - P0) + P0;
T = Timed(end) / temp_s;
dt = Ts;

[Time, P_data, dP_data, ddP_data] = simulateGMP_nDoF(gmp, P0, Pg, T, dt);
toc

%% Reference trajectory (scaled)
Timed = Timed / temp_s;
Pd2_data = spat_s.*( Pd_data-P0 ) + P0;
dPd2_data = spat_s.*dPd_data*temp_s;
ddPd2_data = spat_s.*ddPd_data*temp_s^2;

%% Plot results
for i=1:3
    figure;
    subplot(3,1,1);
    hold on;
    plot(Time, P_data(i,:), 'LineWidth',2.0 , 'Color','blue');
    plot(Timed, Pd2_data(i,:), 'LineWidth',2.0, 'LineStyle',':', 'Color','magenta');
    ylabel('pos [$m$]', 'interpreter','latex', 'fontsize',15);
    title(['temporal scale: $' num2str(temp_s) '$     ,     spatial scale: $' num2str(spat_s(1)) ', ' num2str(spat_s(2)), ', ' num2str(spat_s(3)) '$'], 'interpreter','latex', 'fontsize',18);
    legend({'sim','$k_s$*demo'}, 'interpreter','latex', 'fontsize',15);

    axis tight;
    hold off;

    subplot(3,1,2);
    hold on;
    plot(Time, dP_data(i,:), 'LineWidth',2.0, 'Color','blue');
    plot(Timed, dPd2_data(i,:), 'LineWidth',2.0, 'LineStyle',':', 'Color','magenta');
    ylabel('vel [$m/s$]', 'interpreter','latex', 'fontsize',15);
    axis tight;
    hold off;

    subplot(3,1,3);
    hold on;
    plot(Time, ddP_data(i,:), 'LineWidth',2.0, 'Color','blue');
    plot(Timed, ddPd2_data(i,:), 'LineWidth',2.0, 'LineStyle',':', 'Color','magenta');
    ylabel('accel [$m/s^2$]', 'interpreter','latex', 'fontsize',15);
    axis tight;
    hold off;
end


figure;
hold on;
plot3(P_data(1,:), P_data(2,:), P_data(3,:), 'LineWidth',2, 'LineStyle','-', 'Color','magenta');
plot3(Pd2_data(1,:), Pd2_data(2,:), Pd2_data(3,:), 'LineWidth',2, 'LineStyle',':', 'Color','blue');
plot3(Pd_data(1,:), Pd_data(2,:), Pd_data(3,:), 'LineWidth',2, 'LineStyle',':', 'Color',[0 0.7 0]);
legend({'sim','$k_s$*demo','demo'}, 'interpreter','latex', 'fontsize',15);
xlabel('$x$', 'interpreter','latex', 'fontsize',15);
ylabel('$y$', 'interpreter','latex', 'fontsize',15);
zlabel('$z$', 'interpreter','latex', 'fontsize',15);
hold off;


%% ============================================================
%% ============================================================


function [Time, Y_data, dY_data, ddY_data] = simulateGMP_nDoF(gmp, y0, g, T, dt)
%% Simulates a dmp
% @param[in] gmp: Dim x 1 cell array, where each cell is a 1D GMP.
% @param[in] y0: Dim x 1 vector with the initial position..
% @param[in] g: Dim x 1 vector with the goal/target position.
% @param[in] T: Movement total time duration.
% @param[in] dt: Simulation timestep.
% @param[out] Time: 1 x N rowvector with simulation output timestamps.
% @param[out] Y_data: Dim x N matrix with simulation output positions.
% @param[out] dY_data: Dim x N matrix with simulation output velocities.
% @param[out] ddY_data: Dim x N matrix with simulation output accelerations.
%


%% set initial values
Dim = gmp.numOfDoFs();
ddy = zeros(Dim,1);
dy = zeros(Dim,1);
y = y0;
t = 0.0;
dz = zeros(Dim,1);
z = zeros(Dim,1);

t_end = T;
tau = t_end;
x = 0.0;
x_dot = 1/tau;
s = GMP_phase(x, x_dot, 0);

iters = 0;
Time = [];
Y_data = [];
dY_data = [];
ddY_data = [];
x_data = [];

gmp.setY0(y0);
gmp.setGoal(g);

%% simulate
while (true)

    %% data logging
    Time = [Time t];
    Y_data = [Y_data y];
    dY_data = [dY_data dy];  
    ddY_data = [ddY_data ddy];
    % x_data = [x_data x];

    %% DMP simulation
    y_c = 0.0;
    z_c = 0.0;
    gmp.update(s, y, z, y_c, z_c);
    dy = gmp.getYdot();
    dz = gmp.getZdot();

    yc_dot = 0.0;
    ddy = gmp.getYddot(yc_dot);

    %% Stopping criteria
    if (t>=1.1*t_end) % && norm(y-g)<5e-3 && norm(dy)<5e-3)
        break;
    end

    %% Numerical integration
    iters = iters + 1;
    t = t + dt;
    x = x + x_dot*dt;
    y = y + dy*dt;
    z = z + dz*dt;

    s.x = x;
    
end


end

File: src/gmp_test/src/simulation/nDoF_gmp_test.cpp
#include <cstdlib>
#include <memory>

#include <ros/ros.h>
#include <ros/package.h>

#include <armadillo>

#include <gmp_lib/gmp_lib.h>

#include <gmp_test/utils/utils.h>

using namespace as64_;

std::string train_filename;
std::string results_filename;
std::string train_method;
int N_kernels;
double kernels_std_scaling;
std::string scale_type;
arma::vec wb_normal;
arma::vec Pg;
double T;

bool read_gmp_from_file;
bool write_gmp_to_file;
std::string gmp_filename;

void loadParams();

// ==================================
// ------------   MAIN  -------------
// ==================================

int main(int argc, char **argv)
{
  ros::init(argc, argv, "nDoF_gmp_test_node");

  // =============  Load params  =============
  loadParams();

  // =============  Load train data  =============
  arma::rowvec Timed;
  arma::mat Pd_data, dPd_data, ddPd_data;
  gmp_::FileIO fid(train_filename, gmp_::FileIO::in);
  fid.read("Timed",Timed);
  fid.read("Pd_data",Pd_data);
  fid.read("dPd_data",dPd_data);
  fid.read("ddPd_data",ddPd_data);
  fid.close();

  double Ts = Timed(1) - Timed(0);

  // =============  Create/Train GMP  =============
  gmp_::GMP::Ptr gmp( new gmp_::GMP() );

  unsigned n_dof = Pd_data.n_rows;
  if (read_gmp_from_file)
  {
    gmp_::read(gmp.get(), gmp_filename, "");
    PRINT_INFO_MSG("Loaded GMP from file!\n");
  }
  else
  {
    // initialize and train GMP
    gmp.reset( new gmp_::GMP(n_dof, N_kernels, kernels_std_scaling) );
    Timer::tic();
    arma::vec offline_train_mse;
    PRINT_INFO_MSG("GMP training...\n");
    gmp->train(train_method, Timed/Timed.back(), Pd_data, &offline_train_mse);
    std::cerr << "offline_train_mse = \n" << offline_train_mse << "\n";
    Timer::toc();
  }

  // set scaling type
  gmp_::TrajScale::Ptr traj_sc;
  if (scale_type.compare("prop") == 0) traj_sc.reset( new gmp_::TrajScale_Prop(n_dof) );
  else if (scale_type.compare("rot_min") == 0) traj_sc.reset( new gmp_::TrajScale_Rot_min() );
  else if (scale_type.compare("rot_wb") == 0)
  {
    traj_sc.reset( new gmp_::TrajScale_Rot_wb() );
    dynamic_cast<gmp_::TrajScale_Rot_wb *>(traj_sc.get())->setWorkBenchNormal( {0, 0, 1} );
  }
  else throw_error("Unsupported scale type \"" + scale_type + "\"...\n");

  gmp->setScaleMethod(traj_sc);

  if (write_gmp_to_file)
  {
    gmp_::write(gmp.get(), gmp_filename, "");
    PRINT_INFO_MSG("Wrote GMP to file!\n");
  }

  // gmp->autoRetrain(50, 2, 200, "LWR");

  // =============  GMP simulation  =============
  PRINT_INFO_MSG("GMP simulation...\n");

  Timer::tic();

  int n_data = Timed.size();
  int i_end = n_data - 1;
  arma::vec P0d = Pd_data.col(0);
  arma::vec Pgd = Pd_data.col(i_end);
  arma::vec P0 = P0d;
  // arma::vec Pg = spat_s%(Pgd - P0) + P0;
  // double T = Timed(i_end) / temp_s;
  double dt = Ts;

  arma::rowvec Time;
  arma::mat P_data, dP_data, ddP_data;
  simulateGMP(gmp, P0, Pg, T, dt, Time, P_data, dP_data, ddP_data);

  Timer::toc();

  // This is the groundtruth trajectory that should be produced
  arma::mat Ks = gmp->getScaling();
  double temp_s = Timed.back() / T; // temporal scaling 
  // arma::rowvec Timed2 = Timed / temp_s;
  // arma::mat Pd2_data = Ks*( Pd_data-P0d ) + P0;
  // arma::mat dPd2_data = Ks*dPd_data*temp_s;
  // arma::mat ddPd2_data = Ks*ddPd_data*std::pow(temp_s,2);

  // =============  Write results  =============
  {
    gmp_::FileIO fid(results_filename, gmp_::FileIO::out | gmp_::FileIO::trunc);
    // write the demo data
    fid.write("Timed",Timed);
    fid.write("Pd_data",Pd_data);
    fid.write("dPd_data",dPd_data);
    fid.write("ddPd_data",ddPd_data);
    // write the DMP data
    fid.write("Time",Time);
    fid.write("P_data",P_data);
    fid.write("dP_data",dP_data);
    fid.write("ddP_data",ddP_data);
    // write the spatial scaling matrix and the temporal scaling
    fid.write("Ks",Ks);
    fid.write("temp_s",temp_s);
  }

  PRINT_CONFIRM_MSG("<====  Finished! ====>\n");

  return 0;
}

// ===============================================

void loadParams()
{
  std::string package_path = ros::package::getPath(PROJECT_NAME_) + "/";
  ros::NodeHandle nh("~");
  if (!nh.getParam("train_filename", train_filename)) throw_error("Failed to load param \"train_filename\"...\n");
  train_filename = package_path + train_filename;
  if (!nh.getParam("results_filename", results_filename)) throw_error("Failed to load param \"results_filename\"...\n");
  results_filename = package_path + results_filename;
  if (!nh.getParam("train_method", train_method)) throw_error("Failed to load param \"train_method\"...\n");
  if (!nh.getParam("N_kernels", N_kernels)) throw_error("Failed to load param \"N_kernels\"...\n");
  if (!nh.getParam("kernels_std_scaling", kernels_std_scaling)) throw_error("Failed to load param \"kernels_std_scaling\"...\n");
  if (!nh.getParam("scale_type", scale_type)) throw_error("Failed to load param \"scale_type\"...\n");
  std::vector<double> wb_normal_temp;
  if (!nh.getParam("wb_normal", wb_normal_temp)) throw_error("Failed to load param \"wb_normal\"...\n");
  wb_normal = wb_normal_temp;

  std::vector<double> Pg_;
  if (!nh.getParam("Pg", Pg_)) throw_error("Failed to load param \"Pg\"...\n");
  Pg = Pg_;
  if (!nh.getParam("T", T)) throw_error("Failed to load param \"T\"...\n");

  if (!nh.getParam("read_gmp_from_file", read_gmp_from_file)) throw_error("Failed to load param \"read_gmp_from_file\"...\n");
  if (!nh.getParam("write_gmp_to_file", write_gmp_to_file)) throw_error("Failed to load param \"write_gmp_to_file\"...\n");
  if ((read_gmp_from_file || write_gmp_to_file) && !nh.getParam("gmp_filename", gmp_filename))
    throw_error("Failed to load param \"gmp_filename\"...\n");
  gmp_filename = package_path + gmp_filename;
}
